[RFC] Demux blob

juimonen opened the pull request:
This is a RFC for making mux/demux routing matrix configuration easier as the routing matrix is now "visualized" in m4.

juimonen commented:
see also this: https://github.com/thesofproject/sof-docs/pull/203

@akloniex FYI, would appreciate your comments.

@lgirdwood Thinking and trying out different ways of setting the parameters in mux demux, I'm hesitant to use the enum controls, as at least for me they seems very confusing and we loose some of the demux/mux functionality. So I would go for more "visualized" approach in topology and still try to make the command line generator better...

xiulipan commented on tools/topology/sof/pipe-volume-demux-playback.m4:
```diff
@@ -22,8 +22,29 @@ include(`muxdemux.m4')
+			     `BITS_TO_BYTE(0, 0, 0 ,0 ,0 ,0 ,1 ,0)',
+			     `BITS_TO_BYTE(0, 0, 0 ,0 ,0 ,0 ,0 ,1)')')
+
+define(matrix2, `ROUTE_MATRIX(5, 2,
```
can we make this 5 as an DEFINE value here? So we can override it with some define when include this file.
Same for the 2 here, it should use channel variable.

juimonen replied:
@xiuli Artur's PR #2600 will make these fields obsolete, so we should not anymore need to define these at all. I need to remove these after that PR is merged.

juimonen replied:
moved the demux configuration to upper level so pipiline_id's are visible in the same file. You could also use defines. Generally I'm not huge fan of having defines in upper level topology files and using them in lower level building blocks. Makes understanding the structure really difficult...

xiulipan commented on tools/topology/sof/pipe-volume-demux-playback.m4:
```diff
@@ -22,8 +22,29 @@ include(`muxdemux.m4')
-# Use default parameters
-include(`demux_coef_default.m4')
+dnl pipeline_id, channels, routing_matrix_rows
+define(matrix1, `ROUTE_MATRIX(PIPELINE_ID, 2,
```
can we use the channel define ?
```suggestion
define(matrix1, `ROUTE_MATRIX(PIPELINE_ID, PIPELINE_CHANNELS,
```

juimonen replied:
ok, let me try that.

juimonen replied:
removed

xiulipan commented on tools/topology/sof/pipe-volume-demux-playback.m4:
```diff
@@ -22,8 +22,29 @@ include(`muxdemux.m4')
+			     `BITS_TO_BYTE(0, 0, 0 ,0 ,0 ,0 ,0 ,1)')')
+
+dnl frame_format, num_channels, num_streams, route_matrix
+MUXDEMUX_CONFIG(demux_priv, 2, 2, 2, LIST(`	', `matrix1,', `matrix2'))
```
```suggestion
MUXDEMUX_CONFIG(demux_priv, FMT(PIPELINE_FORMAT), PIPELINE_CHANNELS, 2, LIST(`	', `matrix1,', `matrix2'))
```
you will need a function to convert the format in you define in mux.h

juimonen replied:
I think artur's PR is also removing this field also. We should have all pipeline params come from parameter propagation in pipelines. Only valid field would be the number of streams.

akloniex commented:
@juimonen I'm currently working on few mux/demux optimizations.
Few things up front:
- frame format in configuration blob does not have to be valid ATM - pipeline params propagation takes care of that just fine
- number of channels per stream _will not_ need to be vaild after my PR (WIP, just need to split it into logical commits) - same as above, pipeline params propagation does the trick

I'll send the PR today, and if we wish, we might go one step further and remove these fields from the interface altogether (for now I'll leave them marked as reserved). That would obviously cause backward incompatibility, but would cleanup the interface - I will not judge which way to go.

akloniex commented:
@juimonen Please take a look at https://github.com/thesofproject/sof/pull/2600
As I've mentioned before, few fields from the interface are obsolete, which would greatly simplify configuration of MUX/DEMUX. Now the configuration boils down to the routing table and pipe ID per stream configured.

lgirdwood commented:
> @juimonen Please take a look at #2600
> As I've mentioned before, few fields from the interface are obsolete, which would greatly simplify configuration of MUX/DEMUX. Now the configuration boils down to the routing table and pipe ID per stream configured.

@akloniex @juimonen can we make these fields "reserved" in the IPC if they have never been used. If they have been used in the past then please add a "_deprecated" suffix and a cooment stating last supported ABI version.

plbossart commented on tools/topology/m4/muxdemux.m4:
```diff
@@ -2,6 +2,24 @@ divert(-1)
+define(stream_data_size, 16)
+define(mux_config_size, 8)
+
+define(ROUTE_MATRIX, `PRINT_BYTES_4($1),PRINT_BYTE($2),PRINT_BYTE($3),PRINT_BYTE($4),PRINT_BYTE($5),PRINT_BYTE($6),PRINT_BYTE($7),PRINT_BYTE($8),PRINT_BYTE($9),PRINT_BYTE($10),PRINT_BYTE(0),PRINT_BYTE(0),PRINT_BYTE(0)')dnl
```
wouldn't hurt to add a comment on the representation, I don't have a clue why the first part it BYTES_4 and we have three PRINT_BYTE(0) at the end?

juimonen replied:
ok

plbossart commented on tools/topology/m4/muxdemux.m4:
```diff
@@ -2,6 +2,24 @@ divert(-1)
+       `PRINT_BYTES_4(0),PRINT_BYTES_4(0),'
+       `PRINT_BYTES_2($2),PRINT_BYTES_2($3),PRINT_BYTES_2($4),PRINT_BYTES_2(0),'
+       `$5'`"'
+       `}'
```
same, is there a description of the representation?

juimonen replied:
ok, I tried to add the reference to the actual struct. and comment that reserved members are zeroed.

plbossart commented on tools/topology/m4/muxdemux.m4:
```diff
@@ -2,6 +2,24 @@ divert(-1)
 
 dnl Define macro for demux widget
 
+define(comp_type, 0)
```
what is 0?

juimonen replied:
this is not used for anything and will be removed in #2600, so I will remove it also

plbossart commented on tools/topology/m4/muxdemux.m4:
```diff
@@ -2,6 +2,24 @@ divert(-1)
 
+define(comp_type, 0)
+define(sof_magic, 0x00464F53)
+define(stream_data_size, 16)
```
are we limited to 16 bits audio data?

juimonen replied:
oh this/these are the struct sizes needed for the blob total size, maybe better naming should be used..

plbossart commented on tools/topology/m4/muxdemux.m4:
```diff
@@ -2,6 +2,24 @@ divert(-1)
+define(comp_type, 0)
+define(sof_magic, 0x00464F53)
+define(stream_data_size, 16)
+define(mux_config_size, 8)
```
8 what? unit?

juimonen replied:
again struct size, so needs to be more clear

plbossart commented on tools/topology/sof/pipe-volume-demux-playback.m4:
```diff
@@ -22,8 +22,29 @@ include(`muxdemux.m4')
+			     `BITS_TO_BYTE(0, 0, 0 ,0 ,1 ,0 ,0 ,0)',
+			     `BITS_TO_BYTE(0, 0, 0 ,0 ,0 ,1 ,0 ,0)',
+			     `BITS_TO_BYTE(0, 0, 0 ,0 ,0 ,0 ,1 ,0)',
+			     `BITS_TO_BYTE(0, 0, 0 ,0 ,0 ,0 ,0 ,1)')')
```
can you explain what those matrices do? not use if this is a passthrough or a splitter?

juimonen replied:
https://github.com/thesofproject/sof-docs/pull/203 hope this helps, would be good if you have time to browse it through at some point...

plbossart commented on tools/topology/sof/pipe-volume-demux-playback.m4:
```diff
@@ -32,7 +53,7 @@ C_CONTROLBYTES(DEMUX, PIPELINE_ID,
 	CONTROLBYTES_MAX(, 304),
 	,
-	DEMUX_priv)
+	demux_priv)
```
where was DEMUX_priv defined? Should it be removed? Or does this change below in another patch?

juimonen replied:
yes it should be removed and I'll make a follow up patch for it

juimonen commented on tools/topology/m4/utils.m4:
```diff
@@ -114,5 +114,17 @@ define(`VIRTUAL_WIDGET',
+define(`SOF_MAJOR', esyscmd(`grep "#define SOF_ABI_MAJOR " ../../src/include/kernel/abi.h | grep -E ".[[:digit:]]$" -o'))dnl
+define(`SOF_MINOR', esyscmd(`grep "#define SOF_ABI_MINOR " ../../src/include/kernel/abi.h | grep -E ".[[:digit:]]$" -o'))dnl
+define(`SOF_PATCH', esyscmd(`grep "#define SOF_ABI_PATCH " ../../src/include/kernel/abi.h | grep -E ".[[:digit:]]$" -o'))dnl
+define(`SOF_MAJOR_SHIFT', esyscmd(`grep "#define SOF_ABI_MAJOR_SHIFT" ../../src/include/kernel/abi.h | grep -E ".[[:digit:]]$" -o'))dnl
```
yes

juimonen commented:
@xiulipan @lgirdwood @plbossart @akloniex updated.

- cut the m4 macros to 80 chars (or at least shorter than before), there are some issues always as I try to make the "bytes" blob look nice in .conf
- more comments to blob generation, 0 bit fields are mostly for reserved members in structs
- move blob generation to top level, as it is otherwise really hard to see where the pipeline_id's are coming, hope we get rid of this in Artur's PR. This helps also with differentiating the configuration
- remove unused default demux blob and pipeline
- ABI_VER macros should now work with CI, new define SOFABIPATH needs to be used for command line demux topology hacking.

juimonen commented:
And finally, this is PR is made now so that it should be applied only after #2600, so it takes into account the changed demux structs.

lgirdwood commented on tools/topology/m4/muxdemux.m4:
```diff
@@ -2,6 +2,35 @@ divert(-1)
 dnl Define macro for demux widget
 
+dnl Hard coded values for mux/demux config blob
+define(sof_magic, 0x00464F53)
```
Need to namespace these with mux_

juimonen replied:
ok

bardliao commented on tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4:
```diff
@@ -18,6 +19,31 @@ include(`platform/intel/'PLATFORM`.m4')
+			     `BITS_TO_BYTE(0, 0, 0 ,0 ,0 ,0 ,0 ,1)')')
+
+# name, num_streams, route_matrix list
+MUXDEMUX_CONFIG(demux_priv, 2, LIST(`	', `matrix1,', `matrix2'))
```
This is only for 2 rt1308 codecs case. So you will need
```patch
diff --git a/tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4 b/tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4
index f42bf5101..c5f68bf5b 100644
--- a/tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4
+++ b/tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4
@@ -19,6 +19,8 @@ include(`platform/intel/'PLATFORM`.m4')
 
 DEBUG_START
 
+ifdef(`MONO', `',
+`
 # Configure demux
 # name, pipeline_id, routing_matrix_rows
 define(matrix1, `ROUTE_MATRIX(3,
@@ -43,6 +45,7 @@ define(matrix2, `ROUTE_MATRIX(4,
 
 # name, num_streams, route_matrix list
 MUXDEMUX_CONFIG(demux_priv, 2, LIST(`	', `matrix1,', `matrix2'))
+')
 
 #
 # Define the pipelines
```

juimonen replied:
@bardliao I'm having trouble to compile to file with mono flag... shouldn't the demux pipe addition be also behind MONO?

juimonen replied:
@bardliao oh that was actually fixed in master, had old version

bardliao commented:
I tested it on CML-H with rt711+rt1308+rt715. But somehow there is no output from both speakers.

juimonen commented:
@bardliao so this latest version would need #2600 as under it. So 2 things:
- Can you make the device work with only #2600 without this PR?
- If you can, problem is in the blob in this PR, so could you then just make the conf file from topology with and without #2600 so it would be easy to compare....?
- if already #2600 causes issues, that needs to be debugged first

bardliao commented:
> * Can you make the device work with only #2600 without this PR?

@juimonen No, I will see DSP Panic if I apply #2600. I don't know what did I miss. I will re-test it after #2600 is merged.

juimonen commented on tools/topology/m4/utils.m4:
```diff
@@ -114,5 +114,43 @@ define(`VIRTUAL_WIDGET',
+define(`PRINT_BYTE',
+`format(`0x%02x', eval(($1)&0xFF))')dnl
+
+define(`BITS_TO_BYTE',
```
ok

juimonen commented on tools/topology/m4/utils.m4:
```diff
@@ -114,5 +114,43 @@ define(`VIRTUAL_WIDGET',
 `       no_pm "true"'
 `}', `fatal_error(`Invalid parameters ($#) to VIRTUAL_WIDGET')')')
 
+ifdef(`SOFABIPATH', `define(`ABI_PATH', SOFABIPATH)', `define(`ABI_PATH', ../../../src/include/kernel/abi.h)')
```
The path is given to command line system call egrep, so I don't know how to forward it from m4 include... anyway I changed the definition. As the SOF_ABI_VERSION is composed by cmake, at least CI/Docker will find it (when building tools). You can give it from command line now as -DSOF_ABI_FILE= if you want to compile individual demux topology file by hand. Otherwise the macro is empty.

juimonen commented:
@xiulipan @lgirdwood @plbossart @akloniex @bardliao updated.
- commented byte macros
- changed the abi version generation macro
- added mux -prefix to mux hard coded value 
- added MONO ifdefs around icl topology

I'm not sure what to add for the matrices as comments? all matrices are diagonal, so they will copy "channel by channel", so in case of mux input channels are mixed to output channel (all input channel 1's are mixed to output channel 1 etc) in case of demux input channel is copied to output channels (input channel 1 is copied to all output channel 1's etc).

keyonjie commented on tools/topology/sof-apl-demux-pcm512x.m4:
```diff
@@ -19,6 +20,36 @@ include(`platform/intel/bxt.m4')
+			     `BITS_TO_BYTE(0, 0, 0 ,0 ,0 ,1 ,0 ,0)',
+			     `BITS_TO_BYTE(0, 0, 0 ,0 ,0 ,0 ,1 ,0)',
+			     `BITS_TO_BYTE(0, 0, 0 ,0 ,0 ,0 ,0 ,1)')')
+
```
Hi @juimonen since we are aligned that we don't plan to support channel mix features with mux/demux component, how about simplify this like below:
```
define(matrix1, `ROUTE_MATRIX(1, 0, 1, 2, 3, 4, 5, 6, 7)')
```
And change the ROUTE_MATRIX macro definition to be with bit indices inputs.

This will simplify the top level topology files a lot.

juimonen replied:
@keyonjie are we aligned on that matter ? :)

I really would not want to castrate this component... I'm 99% sure we will need the mix feature very soon in some occasion.

keyonjie replied:
@juimonen per my understanding yes. :)

If we are inheriting the MUX/DEMUX concepts from ALSA, which means N->1 or 1->N only, no mixing concept there. But if we want to innovate a more powerful component here (e.g. named shuffler? I actually proposed this to @plbossart and @lgirdwood in 2017), that is another topic.

The pros for the more powerful one is that it is more generic and we don't need to create many components like mixer/mux/channel_selector, just use this unified one with different configure blob.

The cons with it is that the component may run with lower efficiency(we need to check more bit masks and flags) and not so straightforward to be understood, especially the higher MCPS consuming make me nervous and that's why I voted to remove the mixing fancy feature from it.

@lgirdwood please help chime in this please, if mixing is here, then using mapping mechanism to simplify/optimize the mux component may become impossible since it is not 1chan->1chan mapping anymore.

juimonen replied:
@keyon yes I know the alsa mux concept and as you said this component is much more.

This should then be converted as official ASoC mux component, which also forces enum controls to user space.

I actually tried to convert this to ASoC mux at some point but stopped on the way, because this was much more versatile. I also remember talking for example with @singalsuo cases that would need more complicated functioning than simple mux. 
- We could also do following: use same component as "process" component with full API, as here now. 
- Make Asoc mux interface for the same component (restricted usage)

juimonen replied:
@keyon and agree with your observations about performance, the optimizations you mention should be done. However, if you don't know what you are doing, you can do very nasty things with all our components, build really strange pipelines that will consume the whole dsp resources.

Thinking of the interface, I really think the visual "routing matrix" is really the most informative with this kind of component. With enum we would have 2^64 different matrices and it is just an example of a component for which the configuration through ALSA controls is just awkward.

keyonjie replied:
> @Keyon and agree with your observations about performance, the optimizations you mention should be done. However, if you don't know what you are doing, you can do very nasty things with all our components, build really strange pipelines that will consume the whole dsp resources.
> 
> Thinking of the interface, I really think the visual "routing matrix" is really the most informative with this kind of component. With enum we would have 2^64 different matrices and it is just an example of a component for which the configuration through ALSA controls is just awkward.

Agree with that. If we do keep the current mux component design, it is unrealistic to use enum as @lgirdwood mentioned, and your "routing matrix" depicted here is more informative and topology user friendly.

lgirdwood replied:
@juimonen @keyonjie we should be aiming for enums (once mixing is removed). The max numbers of enums/options we are looking at are
1) demux - 8 source channels on 1 stream to map of 8 sink channels on 8 streams i.e. 8 enums of 64 options
2) mux -  8 source channels on 8 streams to 8 sink channels on 1 stream i.e. 8 enums of 64 options max.
In reality we should have less enums and options as 8x8 topologies will be rare.

bardliao commented:
@juimonen Can you rebase to the latest code? There is a conflict on sof-icl-rt711-rt1308-rt715-hdmi.m4. Thanks.

juimonen commented:
@bardliao rebased.

lgirdwood commented:
Jenkins known issues, Travis looks like a false positive.
